src/python/m5/ext/pystats/jsonloader.py

```diff
@@ -0,0 +1,89 @@
+# Copyright (c) 2020 The Regents of The University of California
+# All rights reserved.
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions are
+# met: redistributions of source code must retain the above copyright
+# notice, this list of conditions and the following disclaimer;
+# redistributions in binary form must reproduce the above copyright
+# notice, this list of conditions and the following disclaimer in the
+# documentation and/or other materials provided with the distribution;
+# neither the name of the copyright holders nor the names of its
+# contributors may be used to endorse or promote products derived from
+# this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+# A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+# OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+# SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+# LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+# DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+# THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+from .simstat import SimStat
+from .statistic import Scalar, Distribution, Accumulator
+from .group import Group, Vector
+import json
+
+class JSONLoader(json.JSONDecoder):
+    """
+    Subclass of JSONDecoder that overrides 'object_hook'. Converts JSON object
+    into a SimStat object.
+
+    Usage
+    -----
+    ```
+    from m5.ext.pystats.jsonloader import JSONLoader
+
+    with open(path) as f:
+        simstat_object = json.load(f, cls=JSONLoader)
+    ```
+    """
+
+    def __init__(self):
+        json.JSONDecoder.__init__(self, object_hook=self.__json_to_simstat)
+
+    def __json_to_simstat(self, d: dict) -> SimStat:
+        if 'type' in d:
+            if d['type'] == 'Scalar':
+                d.pop('type', None)
+                return Scalar(**d)
+
+            elif d['type'] == 'Distribution':
+                d.pop('type', None)
+                return Distribution(**d)
+
+            elif d['type'] == 'Accumulator':
+                d.pop('type', None)
+                return Accumulator(**d)
+
+            elif d['type'] == 'Group':
+                return Group(**d)
+
+            elif d['type'] == 'Vector':
+                d.pop('type', None)
+                d.pop('time_conversion', None)
+                return Vector(d)
+        else:
+            return SimStat(**d)
```

Comment:
Can we have some fail-cases in `__json_to_simstat` function? I'm thinking if someone adds a new stat type, this function should throw an "unrecognized type" error.

I'm also a bit uncomfortable that the definition of a SimStat is anything that doesn't have a "type". Is there a way we can improve this somehow? Maybe give SimStat a type of "simstat" ? It would be worrying if someone extended the Python Stats with something that doesn't have a type and they are cast to SimStats (I guess this would throw an error somewhere but it might be hard to diagnose?).

Reply:
Maybe @powerjg could weigh in on this

Reply:
I think this is fine for now. I don't know how better to do this if we're deserializing from json. I think we're going to want to provide simplified output formats (i.e., without the type) and then we'll have to make even more "guesses" with the deserializer.

If you know better ways to solve this problem, Bobby, I'm all ears.